expression/builtin_grouping.go

```diff
@@ -0,0 +1,206 @@
+// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package expression
+
+import (
+	"github.com/gogo/protobuf/proto"
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/chunk"
+	"github.com/pingcap/tipb/go-tipb"
+)
+
+var (
+	_ functionClass = &groupingImplFunctionClass{}
+)
+
+var (
+	_ builtinFunc = &builtinGroupingImplSig{}
+)
+
+type groupingImplFunctionClass struct {
+	baseFunctionClass
+}
+
+func (c *groupingImplFunctionClass) getFunction(ctx sessionctx.Context, args []Expression) (builtinFunc, error) {
+	if err := c.verifyArgs(args); err != nil {
+		return nil, err
+	}
+	argTp := []types.EvalType{types.ETInt}
+	bf, err := newBaseBuiltinFuncWithTp(ctx, c.funcName, args, types.ETInt, argTp...)
+	if err != nil {
+		return nil, err
+	}
+	bf.tp.SetFlen(1)
+	sig := &builtinGroupingImplSig{bf, 0, map[int64]struct{}{}}
+	sig.setPbCode(tipb.ScalarFuncSig_GroupingSig)
+	return sig, nil
+}
+
+// grouping functions called by user is actually executed by this builtinGroupingImplSig.
+// Users will designate a column as parameter to pass into the grouping function, and tidb
+// will rewrite it to convert the parameter to the meta info. Then, tidb will generate grouping id
+// which is a indicator to be calculated with meta info, these grouping id are actually what
+// builtinGroupingImplSig receives.
+type builtinGroupingImplSig struct {
+	baseBuiltinFunc
+
+	// TODO these are two temporary fields for tests
+	mode          tipb.GroupingMode
+	groupingMarks map[int64]struct{}
+}
+
+func (b *builtinGroupingImplSig) SetGroupingMode(mode tipb.GroupingMode) {
+	b.mode = mode
+}
+
+func (b *builtinGroupingImplSig) SetMetaGroupingMarks(groupingMarks map[int64]struct{}) {
+	b.groupingMarks = groupingMarks
+}
+
+func (b *builtinGroupingImplSig) getGroupingMode() tipb.GroupingMode {
+	return b.mode
+}
+
+// metadata returns the metadata of grouping functions
+func (b *builtinGroupingImplSig) metadata() proto.Message {
+	args := &tipb.GroupingFunctionMetadata{}
+	*(args.Mode) = uint32(b.mode)
```

Comment:
Why need to cast to `uint32`, I think both `arg.Mode` and `b.mode` is `tipb.GroupingMode`

Reply:
> Why need to cast to `uint32`, I think both `arg.Mode` and `b.mode` is `tipb.GroupingMode`

data type in tipb in incorrect, I will change the type in tipb first and fix it.